File: src/forager/ui/widgets/download_box.py
"""Compact sidebar download box shown while a download is active."""
from __future__ import annotations
from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import QFrame, QLabel, QVBoxLayout, QHBoxLayout

from forager.ui import style
from forager.ui.theme import C
from forager.ui.pages.downloads import ProgressBar, format_size


class DownloadBox(QFrame):
    """Compact sidebar card shown only while a download is active."""

    clicked = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setCursor(Qt.CursorShape.PointingHandCursor)
        self.setObjectName("downloadBox")
        self.setStyleSheet(
            style.surface_qss(2)
            + f"#downloadBox:hover {{ background-color: {C.COLOR_3}; }}"
        )

        layout = QVBoxLayout(self)
        layout.setContentsMargins(12, 10, 12, 10)
        layout.setSpacing(6)

        top = QHBoxLayout()
        top.setSpacing(6)
        self._name = QLabel("Downloading")
        self._name.setStyleSheet(style.label(self._name, C.TEXT, size=12, weight=600))
        self._percent = QLabel("0%")
        self._percent.setStyleSheet(style.label(self._percent, C.ACCENT_2, size=12, weight=600))
        top.addWidget(self._name)
        top.addStretch(1)
        top.addWidget(self._percent)
        layout.addLayout(top)

        self._bar = ProgressBar(parent=self)
        layout.addWidget(self._bar)

        self._detail = QLabel("")
        self._detail.setStyleSheet(style.label(self._detail, "#b8bcbf", size=11))
        layout.addWidget(self._detail)

        self.hide()

    def mousePressEvent(self, event):
        self.clicked.emit()
        super().mousePressEvent(event)

    def begin(self, name: str) -> None:
        self._name.setText(name)
        self._percent.setText("0%")
        self._bar.set_value(0)
        self._detail.setText("")
        self.show()

    def set_progress(self, progress) -> None:
        percent = f"{progress.percent:.0f}%"
        self._percent.setText(percent)
        self._bar.set_value(progress.percent)
        if progress.stage.lower() == "downloading":
            bits = [f"{format_size(progress.done)} / {format_size(progress.total)}"]
            if progress.speed > 0:
                bits.append(f"{format_size(progress.speed)}/s")
            self._detail.setText(" \u00b7 ".join(bits))
        else:
            self._detail.setText(f"{progress.stage}\u2026")

    def hide_download(self) -> None:
        self.hide()

File: src/ui/widgets/DownloadBox.cpp
#include "ui/widgets/DownloadBox.h"
#include "ui/pages/Downloads.h"
#include "ui/Theme.h"
#include "ui/Style.h"
#include "utils/Format.h"

using namespace theme::C;

DownloadBox::DownloadBox(QWidget* parent)
    : QFrame(parent)
{
    setCursor(Qt::PointingHandCursor);
    setObjectName("downloadBox");
    setStyleSheet(QStringLiteral(
        "QFrame#downloadBox { background-color: %1; border-radius: %2px; }"
        "QFrame#downloadBox:hover { background-color: %3; }")
        .arg(COLOR_2).arg(RADIUS).arg(COLOR_3));

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(12, 10, 12, 10);
    layout->setSpacing(6);

    auto* top = new QHBoxLayout();
    top->setSpacing(6);
    m_name = new QLabel("Downloading");
    style::label(m_name, TEXT, 12, 600);
    m_percent = new QLabel("0%");
    style::label(m_percent, ACCENT_2, 12, 600);
    top->addWidget(m_name);
    top->addStretch(1);
    top->addWidget(m_percent);
    layout->addLayout(top);

    m_bar = new ProgressBar(4, this);
    layout->addWidget(m_bar);

    m_detail = new QLabel();
    style::label(m_detail, TEXT_FAINT, 11);
    layout->addWidget(m_detail);

    hide();
}

void DownloadBox::mousePressEvent(QMouseEvent* event) {
    emit clicked();
    QFrame::mousePressEvent(event);
}

void DownloadBox::begin(const QString& name) {
    m_name->setText(name);
    m_percent->setText("0%");
    m_bar->setValue(0);
    m_detail->clear();
    show();
}

void DownloadBox::setProgress(double percent, const QString& stage, double speed, double done, double total) {
    m_percent->setText(QStringLiteral("%1%").arg(static_cast<int>(percent)));
    m_bar->setValue(percent);
    if (stage.toLower() == "downloading") {
        QStringList bits;
        bits << QStringLiteral("%1 / %2").arg(format::size(done), format::size(total));
        if (speed > 0)
            bits << QStringLiteral("%1/s").arg(format::size(speed));
        m_detail->setText(bits.join(" \u00b7 "));
    } else {
        m_detail->setText(QStringLiteral("%1\u2026").arg(stage));
    }
}

void DownloadBox::hideDownload() {
    hide();
}
